### PR TITLE
[LLVM][TableGen] Change SeachableTableEmitter to use const RecordKeeper

### PR DESCRIPTION
**Pull request by jurahul**
Change SeachableTableEmitter to use const RecordKeeper.
Also change RecordRecTy to use const Record pointers for its classes.

This is a part of effort to have better const correctness in TableGen backends:

https://discourse.llvm.org/t/psa-planned-changes-to-tablegen-getallderiveddefinitions-api-potential-downstream-breakages/81089

### CODE REVIEW

File: llvm/include/llvm/TableGen/Record.h
Comment (arsenm) [inline llvm/include/llvm/TableGen/Record.h:1560]: Is this removed in later patches?

Comment (jurahul) [reply llvm/include/llvm/TableGen/Record.h:1560]: This was needed because the Record * pointers in `RecordRecTy::getClasses` are now const, so this code in TGParser.cpp does not work:

```
      // Add a reference to this field if we know the record class.
      if (TrackReferenceLocs) {
        if (auto *DI = dyn_cast<DefInit>(Result)) {
          DI->getDef()->getValue(FieldName)->addReferenceLoc(FieldNameLoc);
        } else if (auto *TI = dyn_cast<TypedInit>(Result)) {
          if (auto *RecTy = dyn_cast<RecordRecTy>(TI->getType())) {
            for (const Record *R : RecTy->getClasses())
              if (auto *RV = R->getValue(FieldName))
                RV->addReferenceLoc(FieldNameLoc);
          }
        }
      }
```

However, may be its better to localize the const_cast here as opposed to making the member mutable since its just that one place. I'll do that

Comment (jurahul) [reply llvm/include/llvm/TableGen/Record.h]: Done. PTAL

**Review by arsenm — APPROVED** [review]

**Comment by jurahul** [comment]
I have the fix for LIBC build failure in flight. Will commit once its approved (or when checks complete). Pre-commit builds do not seem to build this target.

**Pull request merged**
